qiskit_experiments/analysis/curve_fitting.py

```diff
@@ -0,0 +1,220 @@
+
+    # Compute chi-squared for fit
+    yfits = func(xdata, *popt)
+    chisq = np.mean(((yfits - ydata) / sigma) ** 2)
+
+    # Compute xdata range for fit
+    xdata_range = [min(xdata), max(xdata)]
+
+    result = {
+        "popt": popt,
+        "popt_err": popt_err,
+        "pcov": pcov,
+        "chisq": chisq,
+        "xrange": xdata_range,
+    }
+    # TODO:
+    #  1. Add some basic validation of computer good / bad based on fit result.
+    #  2. Add error handling so if fitting fails we can return an analysis
+    #     result containing this information
+    return AnalysisResult(result)
+
+
+def multi_curve_fit(
+    funcs: List[Callable],
+    xdata: np.ndarray,
+    ydata: np.ndarray,
+    p0: np.ndarray,
+    sigma: Optional[np.ndarray] = None,
+    weights: Optional[np.ndarray] = None,
+    **kwargs,
+):
+    """Use non-linear least squares to fit a list of functions, f_i, to data
+
+    Args:
+        funcs: a list of objective functions with signatures `f_i(x, *params)`.
+        xdata: a 2D array of xdata and function function indexes.
+        ydata: a 1D array of ydata
+        p0: initial guess for optimization parameters.
+        sigma: Optional, a 1D array of standard deviations in ydata.
+        weights: Optional, a 1D list of numeric weights for each function.
+        kwargs: additional kwargs for scipy.optimize.curve_fit.
+
+    Returns:
+        AnalysisResult: result containing `popt` the optimal fit parameters,
+                        `popt_err` the standard error estimates popt,
+                        `pcov` the covariance matrix for the fit,
+                        `chisq` the chi-squared parameter of fit.
+                        `xrange` the range of xdata values used for fit.
+
+    Raises:
+        QiskitError: if input xdata is not 2D.
+    """
+    num_funcs = len(funcs)
+
+    # Get 1D xdata and indices from 2D input xdata
+    xdata = np.asarray(xdata, dtype=float)
+    if xdata.ndim != 2:
+        raise QiskitError("multi_curve_fit requires 2D xdata.")
+    xdata1d, xindex = xdata.T
+
+    # Get positions for indexes data sets
+    idxs = [xindex == i for i in range(num_funcs)]
+
+    # Combine weights and sigma for transformation
+    if weights is None:
+        wsigma = sigma
+    else:
+        wsigma = np.zeros(ydata.size)
+        if sigma is None:
+            for i in range(num_funcs):
+                wsigma[idxs[i]] = 1 / np.sqrt(weights[i])
+        else:
+            for i in range(num_funcs):
+                wsigma[idxs[i]] = sigma[idxs[i]] / np.sqrt(weights[i])
+
+    # Define multi-objective function
+    def f(x, *params):
+        y = np.zeros(x.size)
+        for i in range(num_funcs):
+            xi = x[idxs[i]]
+            yi = funcs[i](xi, *params)
+            y[idxs[i]] = yi
+        return y
+
+    # Run linearized curve_fit
+    analysis_result = curve_fit(f, xdata1d, ydata, p0, sigma=wsigma, **kwargs)
+
+    return analysis_result
+
+
+def curve_fit_data(
```

Comment:
I feel like there is a good chance for confusion between `curve_fit` and `curve_fit_data`. Maybe this could be called `prepare_curve` or `get_curve`?

Reply:
I was thinking of it is formatting the input data for curve fit (hence name). Maybe `curve_data` or `process_curve_data`?

Reply:
Those seem okay, too.